server/src/main/java/org/elasticsearch/action/admin/indices/diskusage/IndexDiskUsageAnalyzer.java

```diff
@@ -0,0 +1,666 @@
+                        sortedSet.lookupOrd(ord);
+                    }
+                    break;
+                default:
+                    assert false : "Unknown docValues type [" + dvType + "]";
+                    throw new IllegalStateException("Unknown docValues type [" + dvType + "]");
+            }
+            final long bytesRead = directory.getBytesRead();
+            if (bytesRead > 0) {
+                stats.addDocValues(fieldInfo.name, bytesRead);
+            }
+        }
+    }
+
+    void analyzePostings(SegmentReader reader, IndexDiskUsageStats stats) throws IOException {
+        // TODO: FieldsReader has stats() which might contain the disk usage infos
+        // Also, can we track the byte reader per field extension to avoid visiting terms multiple times?
+        FieldsProducer postingsReader = reader.getPostingsReader();
+        if (postingsReader == null) {
+            return;
+        }
+        postingsReader = postingsReader.getMergeInstance();
+        PostingsEnum postings = null;
+        for (FieldInfo field : reader.getFieldInfos()) {
+            if (field.getIndexOptions() == IndexOptions.NONE) {
+                continue;
+            }
+            cancellationChecker.checkForCancellation();
+            directory.resetAndTrackBytesRead(BytesReadTrackingMode.SIZE_AND_POSITION);
+            final Terms terms = postingsReader.terms(field.name);
+            if (terms == null) {
+                continue;
+            }
+            // Visit terms index and term dictionary
+            final long termBytes;
+            {
+                final BytesRefIterator termsIterator = terms.iterator();
+                final TermsEnum termsEnum = terms.iterator();
+                BytesRef bytesRef;
+                while ((bytesRef = termsIterator.next()) != null) {
+                    cancellationChecker.logEvent();
+                    final boolean found = termsEnum.seekExact(bytesRef);
+                    assert found;
+                }
+                termBytes = directory.getBytesRead();
+                if (termBytes > 0) {
+                    stats.addTerms(field.name, termBytes);
+                }
+            }
+
+            // Visit posting and skip lists
+            final long postingsBytes;
+            {
+                final TermsEnum termsEnum = terms.iterator();
+                while (termsEnum.next() != null) {
+                    termsEnum.totalTermFreq();
+                    postings = termsEnum.postings(postings, PostingsEnum.NONE);
+                    while (postings.nextDoc() != DocIdSetIterator.NO_MORE_DOCS) {
+                        cancellationChecker.logEvent();
+                        postings.freq();
+                    }
+                    // skip lists
+                    for (long idx = 1; idx <= 8; idx++) {
+                        cancellationChecker.logEvent();
+                        final int skipDocID = Math.toIntExact(idx * reader.maxDoc() / 8);
```

Comment:
smart :) it might not read all skip data, but hopefully most of it!

Reply:
Thanks, I learnt this from CheckIndex.